# TV Script Generation

In this project, you'll generate your own [Seinfeld](https://en.wikipedia.org/wiki/Seinfeld) TV scripts using RNNs.  You'll be using part of the [Seinfeld dataset](https://www.kaggle.com/thec03u5/seinfeld-chronicles#scripts.csv) of scripts from 9 seasons.  The Neural Network you'll build will generate a new ,"fake" TV script, based on patterns it recognizes in this training data.

## Get the Data

The data is already provided for you in `./data/Seinfeld_Scripts.txt` and you're encouraged to open that file and look at the text. 
>* As a first step, we'll load in this data and look at some samples. 
* Then, you'll be tasked with defining and training an RNN to generate a new script!

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# load in data
import helper
data_dir = './data/Seinfeld_Scripts.txt'
text = helper.load_data(data_dir)

## Explore the Data
Play around with `view_line_range` to view different parts of the data. This will give you a sense of the data you'll be working with. You can see, for example, that it is all lowercase text, and each new line of dialogue is separated by a newline character `\n`.

In [2]:
view_line_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in text.split()})))

lines = text.split('\n')
print('Number of lines: {}'.format(len(lines)))
word_count_line = [len(line.split()) for line in lines]
print('Average number of words in each line: {}'.format(np.average(word_count_line)))

print()
print('The lines {} to {}:'.format(*view_line_range))
print('\n'.join(text.split('\n')[view_line_range[0]:view_line_range[1]]))

Dataset Stats
Roughly the number of unique words: 46367
Number of lines: 109233
Average number of words in each line: 5.544240293684143

The lines 0 to 10:
jerry: do you know what this is all about? do you know, why were here? to be out, this is out...and out is one of the single most enjoyable experiences of life. people...did you ever hear people talking about we should go out? this is what theyre talking about...this whole thing, were all out now, no one is home. not one person here is home, were all out! there are people trying to find us, they dont know where we are. (on an imaginary phone) did you ring?, i cant find him. where did he go? he didnt tell me where he was going. he must have gone out. you wanna go out you get ready, you pick out the clothes, right? you take the shower, you get all ready, get the cash, get your friends, the car, the spot, the reservation...then youre standing around, what do you do? you go we gotta be getting back. once youre out, you wanna get back! y

---
## Implement Pre-processing Functions
The first thing to do to any dataset is pre-processing.  Implement the following pre-processing functions below:
- Lookup Table
- Tokenize Punctuation

### Lookup Table
To create a word embedding, you first need to transform the words to ids.  In this function, create two dictionaries:
- Dictionary to go from the words to an id, we'll call `vocab_to_int`
- Dictionary to go from the id to word, we'll call `int_to_vocab`

Return these dictionaries in the following **tuple** `(vocab_to_int, int_to_vocab)`

In [3]:
import problem_unittests as tests
from collections import Counter

def create_lookup_tables(text):
    """
    Create lookup tables for vocabulary
    :param text: The text of tv scripts split into words
    :return: A tuple of dicts (vocab_to_int, int_to_vocab)
    """
    # TODO: Implement Function
    word_counts = Counter(text)
    # Sort words by frequency
    sorted_vocab = sorted(word_counts, key=word_counts.get, reverse=True)
    # Create tokenize dicts
    vocab_to_int = {word:idx for idx,word in enumerate(sorted_vocab)}
    int_to_vocab = {idx:word for word,idx in vocab_to_int.items()}
    #print(vocab_to_int)
    #print(int_to_vocab)
    # return tuple
    return (vocab_to_int, int_to_vocab)


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_create_lookup_tables(create_lookup_tables)

Tests Passed


### Tokenize Punctuation
We'll be splitting the script into a word array using spaces as delimiters.  However, punctuations like periods and exclamation marks can create multiple ids for the same word. For example, "bye" and "bye!" would generate two different word ids.

Implement the function `token_lookup` to return a dict that will be used to tokenize symbols like "!" into "||Exclamation_Mark||".  Create a dictionary for the following symbols where the symbol is the key and value is the token:
- Period ( **.** )
- Comma ( **,** )
- Quotation Mark ( **"** )
- Semicolon ( **;** )
- Exclamation mark ( **!** )
- Question mark ( **?** )
- Left Parentheses ( **(** )
- Right Parentheses ( **)** )
- Dash ( **-** )
- Return ( **\n** )

This dictionary will be used to tokenize the symbols and add the delimiter (space) around it.  This separates each symbols as its own word, making it easier for the neural network to predict the next word. Make sure you don't use a value that could be confused as a word; for example, instead of using the value "dash", try using something like "||dash||".

In [4]:


def token_lookup():
    """
    Generate a dict to turn punctuation into a token.
    :return: Tokenized dictionary where the key is the punctuation and the value is the token
    """
    # TODO: Implement Function
    punctToken = {
        '.':'||PERIOD||',
        ',':'||COMMA||',
        '"':'||QUOTATION||',
        ';':'||SEMICLON||',
        '!':'||EXCLAMATION||',
        '?':'||QUESTION||',
        '(':'||LPARANTHESIS||',
        ')':'||RPARANTHESIS||',
        '-':'||DASH||',
        '\n':'||RETURN||'
    }    
    return punctToken

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_tokenize(token_lookup)
print(token_lookup())

Tests Passed
{'.': '||PERIOD||', ',': '||COMMA||', '"': '||QUOTATION||', ';': '||SEMICLON||', '!': '||EXCLAMATION||', '?': '||QUESTION||', '(': '||LPARANTHESIS||', ')': '||RPARANTHESIS||', '-': '||DASH||', '\n': '||RETURN||'}


## Pre-process all the data and save it

Running the code cell below will pre-process all the data and save it to file. You're encouraged to lok at the code for `preprocess_and_save_data` in the `helpers.py` file to see what it's doing in detail, but you do not need to change this code.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# pre-process training data
helper.preprocess_and_save_data(data_dir, token_lookup, create_lookup_tables)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

int_text, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()

## Build the Neural Network
In this section, you'll build the components necessary to build an RNN by implementing the RNN Module and forward and backpropagation functions.

### Check Access to GPU

In [2]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import torch

# Check for a GPU
train_on_gpu = torch.cuda.is_available()
if not train_on_gpu:
    print('No GPU found. Please use a GPU to train your neural network.')

## Input
Let's start with the preprocessed input data. We'll use [TensorDataset](http://pytorch.org/docs/master/data.html#torch.utils.data.TensorDataset) to provide a known format to our dataset; in combination with [DataLoader](http://pytorch.org/docs/master/data.html#torch.utils.data.DataLoader), it will handle batching, shuffling, and other dataset iteration functions.

You can create data with TensorDataset by passing in feature and target tensors. Then create a DataLoader as usual.
```
data = TensorDataset(feature_tensors, target_tensors)
data_loader = torch.utils.data.DataLoader(data, 
                                          batch_size=batch_size)
```

### Batching
Implement the `batch_data` function to batch `words` data into chunks of size `batch_size` using the `TensorDataset` and `DataLoader` classes.

>You can batch words using the DataLoader, but it will be up to you to create `feature_tensors` and `target_tensors` of the correct size and content for a given `sequence_length`.

For example, say we have these as input:
```
words = [1, 2, 3, 4, 5, 6, 7]
sequence_length = 4
```

Your first `feature_tensor` should contain the values:
```
[1, 2, 3, 4]
```
And the corresponding `target_tensor` should just be the next "word"/tokenized word value:
```
5
```
This should continue with the second `feature_tensor`, `target_tensor` being:
```
[2, 3, 4, 5]  # features
6             # target
```

In [3]:
from torch.utils.data import TensorDataset, DataLoader
import numpy as np
import itertools

def batch_data(words, sequence_length, batch_size):
    """
    Batch the neural network data using DataLoader
    :param words: The word ids of the TV scripts
    :param sequence_length: The sequence length of each batch
    :param batch_size: The size of each batch; the number of sequences in a batch
    :return: DataLoader with batched data
    """
    # Sequence length gives the nº of words for each sequence, that is, nº_seqs = batch_size // seq_length
    # TODO: Implement function
    num_batches = len(words)//batch_size # // returns int number, / returns floating prec
    print('num batches: ', num_batches)
    # only full batches
    words = words[:num_batches*batch_size]
    x_batch, y_batch = [],[]
    
    for batch in range(0,num_batches):
        for idx in range(batch*batch_size, batch*batch_size + batch_size - sequence_length):
            x_batch.append(words[idx:idx+sequence_length])
            y_batch.append(words[idx+sequence_length])
        

    dataset = TensorDataset(torch.from_numpy(np.asarray(x_batch, dtype = int)),
                                            torch.from_numpy(np.asarray(y_batch)))
    data_loader = torch.utils.data.DataLoader(dataset, 
                                          batch_size=batch_size, shuffle = True)
        #x.append()
    # return a dataloader
    return data_loader

# there is no test for this function, but you are encouraged to create
# print statements and tests of your own

### Test your dataloader 

You'll have to modify this code to test a batching function, but it should look fairly similar.

Below, we're generating some test text data and defining a dataloader using the function you defined, above. Then, we are getting some sample batch of inputs `sample_x` and targets `sample_y` from our dataloader.

Your code should return something like the following (likely in a different order, if you shuffled your data):

```
torch.Size([10, 5])
tensor([[ 28,  29,  30,  31,  32],
        [ 21,  22,  23,  24,  25],
        [ 17,  18,  19,  20,  21],
        [ 34,  35,  36,  37,  38],
        [ 11,  12,  13,  14,  15],
        [ 23,  24,  25,  26,  27],
        [  6,   7,   8,   9,  10],
        [ 38,  39,  40,  41,  42],
        [ 25,  26,  27,  28,  29],
        [  7,   8,   9,  10,  11]])

torch.Size([10])
tensor([ 33,  26,  22,  39,  16,  28,  11,  43,  30,  12])
```

### Sizes
Your sample_x should be of size `(batch_size, sequence_length)` or (10, 5) in this case and sample_y should just have one dimension: batch_size (10). 

### Values

You should also notice that the targets, sample_y, are the *next* value in the ordered test_text data. So, for an input sequence `[ 28,  29,  30,  31,  32]` that ends with the value `32`, the corresponding output should be `33`.

In [4]:
# test dataloader

test_text = range(50)
t_loader = batch_data(test_text, sequence_length=5, batch_size=10)

data_iter = iter(t_loader)
sample_x, sample_y = data_iter.next()

print(sample_x.shape)
print(sample_x)
print()
print(sample_y.shape)
print(sample_y)

num batches:  5
torch.Size([10, 5])
tensor([[ 40,  41,  42,  43,  44],
        [ 12,  13,  14,  15,  16],
        [ 13,  14,  15,  16,  17],
        [ 23,  24,  25,  26,  27],
        [ 14,  15,  16,  17,  18],
        [ 21,  22,  23,  24,  25],
        [ 11,  12,  13,  14,  15],
        [ 20,  21,  22,  23,  24],
        [ 44,  45,  46,  47,  48],
        [ 22,  23,  24,  25,  26]])

torch.Size([10])
tensor([ 45,  17,  18,  28,  19,  26,  16,  25,  49,  27])


---
## Build the Neural Network
Implement an RNN using PyTorch's [Module class](http://pytorch.org/docs/master/nn.html#torch.nn.Module). You may choose to use a GRU or an LSTM. To complete the RNN, you'll have to implement the following functions for the class:
 - `__init__` - The initialize function. 
 - `init_hidden` - The initialization function for an LSTM/GRU hidden state
 - `forward` - Forward propagation function.
 
The initialize function should create the layers of the neural network and save them to the class. The forward propagation function will use these layers to run forward propagation and generate an output and a hidden state.

**The output of this model should be the *last* batch of word scores** after a complete sequence has been processed. That is, for each input sequence of words, we only want to output the word scores for a single, most likely, next word.

### Hints

1. Make sure to stack the outputs of the lstm to pass to your fully-connected layer, you can do this with `lstm_output = lstm_output.contiguous().view(-1, self.hidden_dim)`
2. You can get the last batch of word scores by shaping the output of the final, fully-connected layer like so:

```
# reshape into (batch_size, seq_length, output_size)
output = output.view(batch_size, -1, self.output_size)
# get last batch
out = output[:, -1]
```

In [5]:
import torch.nn as nn

class RNN(nn.Module):
    
    def __init__(self, vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5):
        """
        Initialize the PyTorch RNN Module
        :param vocab_size: The number of input dimensions of the neural network (the size of the vocabulary)
        :param output_size: The number of output dimensions of the neural network
        :param embedding_dim: The size of embeddings, should you choose to use them        
        :param hidden_dim: The size of the hidden layer outputs
        :param dropout: dropout to add in between LSTM/GRU layers
        """
        super(RNN, self).__init__()
        # TODO: Implement function
        self.output_size = output_size
        self.n_layers = n_layers
        self.hidden_dim = hidden_dim
        
        # set class variables
        # embedding and LSTM layers
        self.embedding = nn.Embedding(vocab_size, embedding_dim)
        self.lstm = nn.LSTM(embedding_dim, hidden_dim, n_layers,
                           dropout=dropout, batch_first=True)
        
        # define model layers
        # linear layer
        self.fc = nn.Linear(hidden_dim, output_size)
    
    def forward(self, nn_input, hidden):
        """
        Forward propagation of the neural network
        :param nn_input: The input to the neural network
        :param hidden: The hidden state        
        :return: Two Tensors, the output of the neural network and the latest hidden state
        """
        # TODO: Implement function
        batch_size = nn_input.size(0)
        
        # embeddings and lstm_out
        embeds = self.embedding(nn_input)
        lstm_output, hidden = self.lstm(embeds, hidden)
        lstm_output = lstm_output.contiguous().view(-1, self.hidden_dim)
        
        # fully-connected layer
        output = self.fc(lstm_output)
        
        # reshape into (batch_size, seq_length, output_size)
        output = output.view(batch_size, -1, self.output_size)
        
        # get last batch
        out = output[:, -1]

        # return one batch of output word scores and the hidden state
        return out, hidden
    
    
    def init_hidden(self, batch_size):
        '''
        Initialize the hidden state of an LSTM/GRU
        :param batch_size: The batch_size of the hidden state
        :return: hidden state of dims (n_layers, batch_size, hidden_dim)
        '''
        # Implement function
        weight = next(self.parameters()).data
        
        # initialize hidden state with zero weights, and move to GPU if available
        if (train_on_gpu):
            hidden = (weight.new(self.n_layers, batch_size, self.hidden_dim).zero_().cuda(),
                     weight.new(self.n_layers, batch_size, self.hidden_dim).zero_().cuda())
        else:
            hidden = (weight.new(self.n_layers, batch_size, self.hidden_dim).zero_(),
                     weight.new(self.n_layers, batch_size, self.hidden_dim).zero_())
            
        return hidden
    
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_rnn(RNN, train_on_gpu)

Tests Passed


### Define forward and backpropagation

Use the RNN class you implemented to apply forward and back propagation. This function will be called, iteratively, in the training loop as follows:
```
loss = forward_back_prop(decoder, decoder_optimizer, criterion, inp, target)
```

And it should return the average loss over a batch and the hidden state returned by a call to `RNN(inp, hidden)`. Recall that you can get this loss by computing it, as usual, and calling `loss.item()`.

**If a GPU is available, you should move your data to that GPU device, here.**

In [6]:
def forward_back_prop(rnn, optimizer, criterion, inp, target, hidden):
    """
    Forward and backward propagation on the neural network
    :param decoder: The PyTorch Module that holds the neural network
    :param decoder_optimizer: The PyTorch optimizer for the neural network
    :param criterion: The PyTorch loss function
    :param inp: A batch of input to the neural network
    :param target: The target output for the batch of input
    :return: The loss and the latest hidden state Tensor
    """
    
    # TODO: Implement Function
    
    # move data to GPU, if available
    if train_on_gpu:
        inp, target = inp.cuda(), target.cuda()
    
    hidden = tuple([each.data for each in hidden])
    
    # zero accumulated gradients
    rnn.zero_grad()
    
    # get the ouput from the RNN model
    output, hidden = rnn(inp, hidden)
    
    # perform backpropagation and optimization
    # calculate the loss and perform backprop
    loss = criterion(output, target)
    loss.backward()
    
    # 'clip_grad_norm' helps prevent the exploding gradient problem in RNNs / LSTMs
    nn.utils.clip_grad_norm_(rnn.parameters(), 5)
    optimizer.step()
    
    # return the loss over a batch and the hidden state produced by our model
    return loss.item(), hidden



# Note that these tests aren't completely extensive.
# they are here to act as general checks on the expected outputs of your functions
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_forward_back_prop(RNN, forward_back_prop, train_on_gpu)

Tests Passed


## Neural Network Training

With the structure of the network complete and data ready to be fed in the neural network, it's time to train it.

### Train Loop

The training loop is implemented for you in the `train_decoder` function. This function will train the network over all the batches for the number of epochs given. The model progress will be shown every number of batches. This number is set with the `show_every_n_batches` parameter. You'll set this parameter along with other parameters in the next section.

In [7]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""

def train_rnn(rnn, batch_size, optimizer, criterion, n_epochs, show_every_n_batches=100):
    batch_losses = []
    
    rnn.train()

    print("Training for %d epoch(s)..." % n_epochs)
    for epoch_i in range(1, n_epochs + 1):
        
        # initialize hidden state
        hidden = rnn.init_hidden(batch_size)
        
        for batch_i, (inputs, labels) in enumerate(train_loader, 1):
            
            # make sure you iterate over completely full batches, only
            n_batches = len(train_loader.dataset)//batch_size
            if(batch_i > n_batches):
                break
            
            # forward, back prop
            loss, hidden = forward_back_prop(rnn, optimizer, criterion, inputs, labels, hidden)          
            # record loss
            batch_losses.append(loss)

            # printing loss stats
            if batch_i % show_every_n_batches == 0:
                print('Epoch: {:>4}/{:<4}  Loss: {}\n'.format(
                    epoch_i, n_epochs, np.average(batch_losses)))
                batch_losses = []

    # returns a trained rnn
    return rnn

### Hyperparameters

Set and train the neural network with the following parameters:
- Set `sequence_length` to the length of a sequence.
- Set `batch_size` to the batch size.
- Set `num_epochs` to the number of epochs to train for.
- Set `learning_rate` to the learning rate for an Adam optimizer.
- Set `vocab_size` to the number of uniqe tokens in our vocabulary.
- Set `output_size` to the desired size of the output.
- Set `embedding_dim` to the embedding dimension; smaller than the vocab_size.
- Set `hidden_dim` to the hidden dimension of your RNN.
- Set `n_layers` to the number of layers/cells in your RNN.
- Set `show_every_n_batches` to the number of batches at which the neural network should print progress.

If the network isn't getting the desired results, tweak these parameters and/or the layers in the `RNN` class.

In [19]:
# Data params
# Sequence Length
sequence_length =  10 # of words in a sequence
# Batch Size
batch_size = 64

# data loader - do not change
train_loader = batch_data(int_text, sequence_length, batch_size)

num batches:  13939


In [20]:
# Training parameters
# Number of Epochs
num_epochs = 10
# Learning Rate
learning_rate = 0.01

# Model parameters
# Vocab size
vocab_size = len(vocab_to_int)
# Output size
output_size = vocab_size # We are looking for the most suitable word, so all words in the dict should be considered
# Embedding Dimension
embedding_dim = 100
# Hidden Dimension
hidden_dim = 128
# Number of RNN Layers
n_layers = 3

# Show stats for every n number of batches
show_every_n_batches = 500

### Train
In the next cell, you'll train the neural network on the pre-processed data.  If you have a hard time getting a good loss, you may consider changing your hyperparameters. In general, you may get better results with larger hidden and n_layer dimensions, but larger models take a longer time to train. 
> **You should aim for a loss less than 3.5.** 

You should also experiment with different sequence lengths, which determine the size of the long range dependencies that a model can learn.

In [21]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from workspace_utils import active_session # To avoid workspace going to sleep with active process

with active_session():
    # create model and move to gpu if available
    rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5)
    if train_on_gpu:
        rnn.cuda()

    # defining loss and optimization functions for training
    optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
    criterion = nn.CrossEntropyLoss()

    # training the model
    trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

    # saving the trained model
    helper.save_model('./save/trained_rnn', trained_rnn)
    print('Model Trained and Saved')

Training for 10 epoch(s)...
Epoch:    1/10    Loss: 6.5355831832885745

Epoch:    1/10    Loss: 6.052312708854675

Epoch:    1/10    Loss: 5.921308146476745

Epoch:    1/10    Loss: 5.877559803962708

Epoch:    1/10    Loss: 5.897878034591675

Epoch:    1/10    Loss: 5.931645053863526

Epoch:    1/10    Loss: 5.956887748718262

Epoch:    1/10    Loss: 5.919408869743347

Epoch:    1/10    Loss: 5.919951305389405

Epoch:    1/10    Loss: 5.8820667676925655

Epoch:    1/10    Loss: 5.932708903312683

Epoch:    1/10    Loss: 5.927751152038574

Epoch:    1/10    Loss: 5.927239100456238

Epoch:    1/10    Loss: 5.871503072738648

Epoch:    1/10    Loss: 5.899904242515564

Epoch:    1/10    Loss: 5.884672795295716

Epoch:    1/10    Loss: 5.88251244354248

Epoch:    1/10    Loss: 5.8678835582733155

Epoch:    1/10    Loss: 5.915889150619507

Epoch:    1/10    Loss: 5.857193439483643

Epoch:    1/10    Loss: 5.897086679458618

Epoch:    1/10    Loss: 5.8696851024627685

Epoch:    1/10    Loss:

/opt/conda/lib/python3.6/site-packages/torch/serialization.py:193: UserWarning: Couldn't retrieve source code for container of type RNN. It won't be checked for correctness upon loading.
  "type " + obj.__name__ + ". It won't be checked "


Model Trained and Saved


In [22]:
# 2nd attempt

# Data params
# Sequence Length
sequence_length =  10 # of words in a sequence
# Batch Size
batch_size = 64

# data loader - do not change
train_loader = batch_data(int_text, sequence_length, batch_size)

# Training parameters
# Number of Epochs
num_epochs = 10
# Learning Rate
learning_rate = 0.005

# Model parameters
# Vocab size
vocab_size = len(vocab_to_int)
# Output size
output_size = vocab_size # We are looking for the most suitable word, so all words in the dict should be considered
# Embedding Dimension
embedding_dim = 100
# Hidden Dimension
hidden_dim = 128
# Number of RNN Layers
n_layers = 3

# Show stats for every n number of batches
show_every_n_batches = 500

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from workspace_utils import active_session # To avoid workspace going to sleep with active process

with active_session():
    # create model and move to gpu if available
    rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5)
    if train_on_gpu:
        rnn.cuda()

    # defining loss and optimization functions for training
    optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
    criterion = nn.CrossEntropyLoss()

    # training the model
    trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

    # saving the trained model
    helper.save_model('./save/trained_rnn', trained_rnn)
    print('Model Trained and Saved')

num batches:  13939
Training for 10 epoch(s)...
Epoch:    1/10    Loss: 6.3116416215896605

Epoch:    1/10    Loss: 5.867983847618103

Epoch:    1/10    Loss: 5.853555173873901

Epoch:    1/10    Loss: 5.864017168998719

Epoch:    1/10    Loss: 5.868488728523254

Epoch:    1/10    Loss: 5.918350551605225

Epoch:    1/10    Loss: 5.9241563510894775

Epoch:    1/10    Loss: 5.887285603523255

Epoch:    1/10    Loss: 5.9299896011352535

Epoch:    1/10    Loss: 5.960600452423096

Epoch:    1/10    Loss: 5.949570366859436

Epoch:    1/10    Loss: 5.978710736274719

Epoch:    1/10    Loss: 5.940274523735046

Epoch:    1/10    Loss: 5.977035046577454

Epoch:    1/10    Loss: 5.917114979743958

Epoch:    1/10    Loss: 5.943578036308288

Epoch:    1/10    Loss: 5.904761220932007

Epoch:    1/10    Loss: 5.910250395774841

Epoch:    1/10    Loss: 5.755046783447265

Epoch:    1/10    Loss: 5.614634372711182

Epoch:    1/10    Loss: 5.589753030776977

Epoch:    1/10    Loss: 5.525954468727112

Epo

KeyboardInterrupt: 

In [24]:
#3rd attempt

# Data params
# Sequence Length
sequence_length =  10 # of words in a sequence
# Batch Size
batch_size = 64

# data loader - do not change
train_loader = batch_data(int_text, sequence_length, batch_size)

# Training parameters
# Number of Epochs
num_epochs = 10
# Learning Rate
learning_rate = 0.001

# Model parameters
# Vocab size
vocab_size = len(vocab_to_int)
# Output size
output_size = vocab_size # We are looking for the most suitable word, so all words in the dict should be considered
# Embedding Dimension
embedding_dim = 100
# Hidden Dimension
hidden_dim = 128
# Number of RNN Layers
n_layers = 3

# Show stats for every n number of batches
show_every_n_batches = 2000

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from workspace_utils import active_session # To avoid workspace going to sleep with active process

with active_session():
    # create model and move to gpu if available
    rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5)
    if train_on_gpu:
        rnn.cuda()

    # defining loss and optimization functions for training
    optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
    criterion = nn.CrossEntropyLoss()

    # training the model
    trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

    # saving the trained model
    helper.save_model('./save/trained_rnn_3th', trained_rnn)
    print('Model Trained and Saved')

num batches:  13939
Training for 10 epoch(s)...
Epoch:    1/10    Loss: 5.400202939271927

Epoch:    1/10    Loss: 4.780879750609398

Epoch:    1/10    Loss: 4.6169420838356015

Epoch:    1/10    Loss: 4.530167439699173

Epoch:    1/10    Loss: 4.455859003186226

Epoch:    2/10    Loss: 4.342307870801983

Epoch:    2/10    Loss: 4.232508268356323

Epoch:    2/10    Loss: 4.241450154185295

Epoch:    2/10    Loss: 4.231174311280251

Epoch:    2/10    Loss: 4.210577802181244

Epoch:    3/10    Loss: 4.151128275855302

Epoch:    3/10    Loss: 4.072905858516693

Epoch:    3/10    Loss: 4.095950795769691

Epoch:    3/10    Loss: 4.104598906159401

Epoch:    3/10    Loss: 4.107317559242248

Epoch:    4/10    Loss: 4.042879832041324

Epoch:    4/10    Loss: 3.9699045951366423

Epoch:    4/10    Loss: 4.018599171161652

Epoch:    4/10    Loss: 4.021038573026657

Epoch:    4/10    Loss: 4.033647597789765

Epoch:    5/10    Loss: 3.9743699622388786

Epoch:    5/10    Loss: 3.920135457158089

Epo

/opt/conda/lib/python3.6/site-packages/torch/serialization.py:193: UserWarning: Couldn't retrieve source code for container of type RNN. It won't be checked for correctness upon loading.
  "type " + obj.__name__ + ". It won't be checked "


In [25]:
# 4th attempt

# Data params
# Sequence Length
sequence_length =  10 # of words in a sequence
# Batch Size
batch_size = 64

# data loader - do not change
train_loader = batch_data(int_text, sequence_length, batch_size)

# Training parameters
# Number of Epochs
num_epochs = 10
# Learning Rate
learning_rate = 0.001

# Model parameters
# Vocab size
vocab_size = len(vocab_to_int)
# Output size
output_size = vocab_size # We are looking for the most suitable word, so all words in the dict should be considered
# Embedding Dimension
embedding_dim = 350
# Hidden Dimension
hidden_dim = 128
# Number of RNN Layers
n_layers = 3

# Show stats for every n number of batches
show_every_n_batches = 2000

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from workspace_utils import active_session # To avoid workspace going to sleep with active process

with active_session():
    # create model and move to gpu if available
    rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5)
    if train_on_gpu:
        rnn.cuda()

    # defining loss and optimization functions for training
    optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
    criterion = nn.CrossEntropyLoss()

    # training the model
    trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

    # saving the trained model
    helper.save_model('./save/trained_rnn_4th', trained_rnn)
    print('Model Trained and Saved')

num batches:  13939
Training for 10 epoch(s)...
Epoch:    1/10    Loss: 5.334638746023178

Epoch:    1/10    Loss: 4.741686530709266

Epoch:    1/10    Loss: 4.559328132271767

Epoch:    1/10    Loss: 4.497009722471237

Epoch:    1/10    Loss: 4.423377056360245

Epoch:    2/10    Loss: 4.291364534203596

Epoch:    2/10    Loss: 4.184394553542137

Epoch:    2/10    Loss: 4.194121618509293

Epoch:    2/10    Loss: 4.19346511721611

Epoch:    2/10    Loss: 4.1834608396291735

Epoch:    3/10    Loss: 4.124486302226441

Epoch:    3/10    Loss: 4.040901533961296

Epoch:    3/10    Loss: 4.038749930739403

Epoch:    3/10    Loss: 4.061933246374131

Epoch:    3/10    Loss: 4.071794291019439

Epoch:    4/10    Loss: 4.009643884708006

Epoch:    4/10    Loss: 3.9531415988206864

Epoch:    4/10    Loss: 3.9607367449998856

Epoch:    4/10    Loss: 4.006284549474716

Epoch:    4/10    Loss: 4.005374879002571

Epoch:    5/10    Loss: 3.944058444835315

Epoch:    5/10    Loss: 3.900747777700424

Epoc

/opt/conda/lib/python3.6/site-packages/torch/serialization.py:193: UserWarning: Couldn't retrieve source code for container of type RNN. It won't be checked for correctness upon loading.
  "type " + obj.__name__ + ". It won't be checked "


In [27]:
#6th attempt

# Data params
# Sequence Length
sequence_length =  10 # of words in a sequence
# Batch Size
batch_size = 64

# data loader - do not change
train_loader = batch_data(int_text, sequence_length, batch_size)

# Training parameters
# Number of Epochs
num_epochs = 10
# Learning Rate
learning_rate = 0.001

# Model parameters
# Vocab size
vocab_size = len(vocab_to_int)
# Output size
output_size = vocab_size # We are looking for the most suitable word, so all words in the dict should be considered
# Embedding Dimension
embedding_dim = 300
# Hidden Dimension
hidden_dim = 256
# Number of RNN Layers
n_layers = 2

# Show stats for every n number of batches
show_every_n_batches = 2000

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from workspace_utils import active_session # To avoid workspace going to sleep with active process

with active_session():
    # create model and move to gpu if available
    rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5)
    if train_on_gpu:
        rnn.cuda()

    # defining loss and optimization functions for training
    optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
    criterion = nn.CrossEntropyLoss()

    # training the model
    trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

    # saving the trained model
    helper.save_model('./save/trained_rnn_6th', trained_rnn)
    print('Model Trained and Saved')

num batches:  13939
Training for 10 epoch(s)...
Epoch:    1/10    Loss: 5.002074063897133

Epoch:    1/10    Loss: 4.474443912982941

Epoch:    1/10    Loss: 4.308876652479172

Epoch:    1/10    Loss: 4.250539356350899

Epoch:    1/10    Loss: 4.196233826518059

Epoch:    2/10    Loss: 4.0405776557755

Epoch:    2/10    Loss: 3.9611507720947268

Epoch:    2/10    Loss: 3.9446490763425825

Epoch:    2/10    Loss: 3.95868026804924

Epoch:    2/10    Loss: 3.9543342052698134

Epoch:    3/10    Loss: 3.85937736706986

Epoch:    3/10    Loss: 3.7948512568473816

Epoch:    3/10    Loss: 3.8048044135570525

Epoch:    3/10    Loss: 3.8056923831701277

Epoch:    3/10    Loss: 3.8437165031433107

Epoch:    4/10    Loss: 3.7557499293981795

Epoch:    4/10    Loss: 3.700048355817795

Epoch:    4/10    Loss: 3.72102422106266

Epoch:    4/10    Loss: 3.7360142867565154

Epoch:    4/10    Loss: 3.75683750808239

Epoch:    5/10    Loss: 3.685009952366146

Epoch:    5/10    Loss: 3.6166541585922243

Ep

/opt/conda/lib/python3.6/site-packages/torch/serialization.py:193: UserWarning: Couldn't retrieve source code for container of type RNN. It won't be checked for correctness upon loading.
  "type " + obj.__name__ + ". It won't be checked "


In [8]:
#7th attempt

# Data params
# Sequence Length
sequence_length =  10 # of words in a sequence
# Batch Size
batch_size = 128

# data loader - do not change
train_loader = batch_data(int_text, sequence_length, batch_size)

# Training parameters
# Number of Epochs
num_epochs = 10
# Learning Rate
learning_rate = 0.001

# Model parameters
# Vocab size
vocab_size = len(vocab_to_int)
# Output size
output_size = vocab_size # We are looking for the most suitable word, so all words in the dict should be considered
# Embedding Dimension
embedding_dim = 300
# Hidden Dimension
hidden_dim = 512
# Number of RNN Layers
n_layers = 2

# Show stats for every n number of batches
show_every_n_batches = 2000

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from workspace_utils import active_session # To avoid workspace going to sleep with active process

with active_session():
    # create model and move to gpu if available
    rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5)
    if train_on_gpu:
        rnn.cuda()

    # defining loss and optimization functions for training
    optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
    criterion = nn.CrossEntropyLoss()

    # training the model
    trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

    # saving the trained model
    helper.save_model('./save/trained_rnn_7th', trained_rnn)
    print('Model Trained and Saved')

num batches:  6969
Training for 10 epoch(s)...
Epoch:    1/10    Loss: 4.688090462327003

Epoch:    1/10    Loss: 4.2023481546640395

Epoch:    1/10    Loss: 4.080353755474091

Epoch:    2/10    Loss: 3.85752870963745

Epoch:    2/10    Loss: 3.810047510266304

Epoch:    2/10    Loss: 3.786196788072586

Epoch:    3/10    Loss: 3.6084599103471233

Epoch:    3/10    Loss: 3.6021103434562685

Epoch:    3/10    Loss: 3.614971899867058

Epoch:    4/10    Loss: 3.453650774833786

Epoch:    4/10    Loss: 3.4349834434986115

Epoch:    4/10    Loss: 3.4784393239021303

Epoch:    5/10    Loss: 3.3150949525361013

Epoch:    5/10    Loss: 3.3246692312955854

Epoch:    5/10    Loss: 3.374431728363037

Epoch:    6/10    Loss: 3.2107415091086535

Epoch:    6/10    Loss: 3.2325219427347185

Epoch:    6/10    Loss: 3.286241559505463

Epoch:    7/10    Loss: 3.1220731140166618

Epoch:    7/10    Loss: 3.1601563557386396

Epoch:    7/10    Loss: 3.2116550174951555

Epoch:    8/10    Loss: 3.0575236986966

/opt/conda/lib/python3.6/site-packages/torch/serialization.py:193: UserWarning: Couldn't retrieve source code for container of type RNN. It won't be checked for correctness upon loading.
  "type " + obj.__name__ + ". It won't be checked "


Model Trained and Saved


### Question: How did you decide on your model hyperparameters? 
For example, did you try different sequence_lengths and find that one size made the model converge faster? What about your hidden_dim and n_layers; how did you decide on those?

#### **Answer:**
In the first attempt I have tried with moderate multiples of 2. In such manner, I've put words to be divided in batches of 64 seqs with seq_length of 10 words. Most senteces have 5-6 words, 10 words are more than enough to guess the whole context of a sentence and generate contextual data. vocab and output size coincide with the dict int we have previously prepared. 64 sequence in a batch offers a a good feedback about the process, allowing to quickly abort it if result is not as expected. Then, I have selected an embedding dim of 100 just to test and 128 as hidden dim (I remember some lesson about encoding-decoding where multiples of 2 were set as preferred to reduce pipeline loss). Finally I set three layers and 10 epochs as are good options to evaluate most neural networks. The learning rate was 0.01. I normally start with 0.01 as it offers a good trade-off between time invested and results as well as allow evaluating the gradient performance. In that case I got 5.12 but the loss was not linearly descending meaning that learning rate should be smaller or complexer the model. 

So, I decreased the learning rate to 0.005 what improved results to 4.14 but still the learning rate was not able to properly approximate the local minima. Hence, I aborted training and decreased the lr even more, down to 0.001.

After 10 epochs the loss is: 3.49. Thus, aiming to achieve better results and evaluate other parameters, the embedding dim have been also tuned in a 5th test. This embedding dim of 350  basis on the lessons, wherein values in the range 200 to 500 are considered adequate for most cases. There we can see that the model converges in a similar fashion. 

A final test has been carried (6th) increasing the lr to 0.002 and hidden_dim to 512 but reducing the layers to 2 and the embedding to 300 which has delivered loss 3.04. Going into test with this model.

------- -----------

Additionally, at the beggining I saw the model quickly converging to values near 0. This unusual overfitting told me that there were some error in the code. Furthermore, when I tested the models the results consisted of repeated sentences without dialogues. This made me go back to the code, after checking the last functions I identified a problem with the get_batch function wherein I was uncorrectly partioning data despite tests and "visual results" seemed correct.

In [26]:
#5th attempt

# Data params
# Sequence Length
sequence_length =  10 # of words in a sequence
# Batch Size
batch_size = 64

# data loader - do not change
train_loader = batch_data(int_text, sequence_length, batch_size)

# Training parameters
# Number of Epochs
num_epochs = 10
# Learning Rate
learning_rate = 0.0001

# Model parameters
# Vocab size
vocab_size = len(vocab_to_int)
# Output size
output_size = vocab_size # We are looking for the most suitable word, so all words in the dict should be considered
# Embedding Dimension
embedding_dim = 350
# Hidden Dimension
hidden_dim = 256
# Number of RNN Layers
n_layers = 3

# Show stats for every n number of batches
show_every_n_batches = 2000

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from workspace_utils import active_session # To avoid workspace going to sleep with active process

with active_session():
    # create model and move to gpu if available
    rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5)
    if train_on_gpu:
        rnn.cuda()

    # defining loss and optimization functions for training
    optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
    criterion = nn.CrossEntropyLoss()

    # training the model
    trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

    # saving the trained model
    helper.save_model('./save/trained_rnn_5th', trained_rnn)
    print('Model Trained and Saved')

num batches:  13939
Training for 10 epoch(s)...
Epoch:    1/10    Loss: 5.265302665352821

Epoch:    1/10    Loss: 4.694258090496064

Epoch:    1/10    Loss: 4.508250891089439

Epoch:    1/10    Loss: 4.4281124167442325

Epoch:    1/10    Loss: 4.3627721126079555

Epoch:    2/10    Loss: 4.213505165845085

Epoch:    2/10    Loss: 4.124460256576538

Epoch:    2/10    Loss: 4.088862963080406

Epoch:    2/10    Loss: 4.102987829089165

Epoch:    2/10    Loss: 4.100464901685715

Epoch:    3/10    Loss: 4.020753105153939

Epoch:    3/10    Loss: 3.9440425677299498

Epoch:    3/10    Loss: 3.934525128006935

Epoch:    3/10    Loss: 3.9427374881505965

Epoch:    3/10    Loss: 3.9792712926864624

Epoch:    4/10    Loss: 3.8863432694546436

Epoch:    4/10    Loss: 3.8110268779993057

Epoch:    4/10    Loss: 3.826832836031914

Epoch:    4/10    Loss: 3.8344099843502044

Epoch:    4/10    Loss: 3.8836550018787386

Epoch:    5/10    Loss: 3.799002191735785

Epoch:    5/10    Loss: 3.72954700255394

/opt/conda/lib/python3.6/site-packages/torch/serialization.py:193: UserWarning: Couldn't retrieve source code for container of type RNN. It won't be checked for correctness upon loading.
  "type " + obj.__name__ + ". It won't be checked "


---
# Checkpoint

After running the above training cell, your model will be saved by name, `trained_rnn`, and if you save your notebook progress, **you can pause here and come back to this code at another time**. You can resume your progress by running the next cell, which will load in our word:id dictionaries _and_ load in your saved model by name!

In [9]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import torch
import helper
import problem_unittests as tests

_, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()
trained_rnn = helper.load_model('./save/trained_rnn_7th')

## Generate TV Script
With the network trained and saved, you'll use it to generate a new, "fake" Seinfeld TV script in this section.

### Generate Text
To generate the text, the network needs to start with a single word and repeat its predictions until it reaches a set length. You'll be using the `generate` function to do this. It takes a word id to start with, `prime_id`, and generates a set length of text, `predict_len`. Also note that it uses topk sampling to introduce some randomness in choosing the most likely next word, given an output set of word scores!

In [10]:
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import torch.nn.functional as F

def generate(rnn, prime_id, int_to_vocab, token_dict, pad_value, predict_len=100):
    """
    Generate text using the neural network
    :param decoder: The PyTorch Module that holds the trained neural network
    :param prime_id: The word id to start the first prediction
    :param int_to_vocab: Dict of word id keys to word values
    :param token_dict: Dict of puncuation tokens keys to puncuation values
    :param pad_value: The value used to pad a sequence
    :param predict_len: The length of text to generate
    :return: The generated text
    """
    rnn.eval()
    
    # create a sequence (batch_size=1) with the prime_id
    current_seq = np.full((1, sequence_length), pad_value)
    current_seq[-1][-1] = prime_id
    predicted = [int_to_vocab[prime_id]]
    
    for _ in range(predict_len):
        if train_on_gpu:
            current_seq = torch.LongTensor(current_seq).cuda()
        else:
            current_seq = torch.LongTensor(current_seq)
        
        # initialize the hidden state
        hidden = rnn.init_hidden(current_seq.size(0))
        
        # get the output of the rnn
        output, _ = rnn(current_seq, hidden)
        
        # get the next word probabilities
        p = F.softmax(output, dim=1).data
        if(train_on_gpu):
            p = p.cpu() # move to cpu
         
        # use top_k sampling to get the index of the next word
        top_k = 5
        p, top_i = p.topk(top_k)
        top_i = top_i.numpy().squeeze()
        
        # select the likely next word index with some element of randomness
        p = p.numpy().squeeze()
        word_i = np.random.choice(top_i, p=p/p.sum())
        
        # retrieve that word from the dictionary
        word = int_to_vocab[word_i]
        predicted.append(word)     
        
        # the generated word becomes the next "current sequence" and the cycle can continue
        current_seq = np.roll(current_seq, -1, 1)
        current_seq[-1][-1] = word_i
    
    gen_sentences = ' '.join(predicted)
    
    # Replace punctuation tokens
    for key, token in token_dict.items():
        ending = ' ' if key in ['\n', '(', '"'] else ''
        gen_sentences = gen_sentences.replace(' ' + token.lower(), key)
    gen_sentences = gen_sentences.replace('\n ', '\n')
    gen_sentences = gen_sentences.replace('( ', '(')
    
    # return all the sentences
    return gen_sentences

### Generate a New Script
It's time to generate the text. Set `gen_length` to the length of TV script you want to generate and set `prime_word` to one of the following to start the prediction:
- "jerry"
- "elaine"
- "george"
- "kramer"

You can set the prime word to _any word_ in our dictionary, but it's best to start with a name for generating a TV script. (You can also start with any other names you find in the original text file!)

In [19]:
# run the cell multiple times to get different results!
gen_length = 600 # modify the length to your preference
prime_word = 'elaine' # name for starting the script

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
pad_word = helper.SPECIAL_WORDS['PADDING']
generated_script = generate(trained_rnn, vocab_to_int[prime_word + ':'], int_to_vocab, token_dict, vocab_to_int[pad_word], gen_length)
print(generated_script)

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:42: UserWarning: RNN module weights are not part of single contiguous chunk of memory. This means they need to be compacted at every call, possibly greatly increasing memory usage. To compact weights again call flatten_parameters().


elaine: fund.

jerry: i mean, you know what i think? i'm going out to get some food.

jerry:(pleading) you know, i think i'm really funny.

kramer: oh, i don't think so.

jerry: so, you want to be a clown.

jerry: oh, you know, i think you can see each other.

george: you know that? it's no big deal.(to elaine) i don't understand.

jerry: you don't know what you're talking about?

kramer: oh, yeah, i'm not happy about it, i gotta go. i can't do it. i can't do it.

kramer: oh, you know that?

elaine:(smiling) oh, you should see that.

elaine: well, it's the damnedest thing i've ever had is a translator for me, i know.(to jerry) i was gonna get the chair.(to elaine) i can't get it. i don't even want to know about it.

kramer:(looking at the doll) yeah.

george:(smiling) no. no, no. no, no. no, no, no, no. no, i mean...

jerry: so what?

jerry: i don't know.

elaine: what?

jerry:(to kramer) i think it's a little effeminate.

jerry: i don't know what the hell is. i'm sorry i'm late, but i

#### Save your favorite scripts

Once you have a script that you like (or find interesting), save it to a text file!

In [16]:
# save script to a text file
f =  open("generated_script_3.txt","w")
f.write(generated_script)
f.close()

# The TV Script is Not Perfect
It's ok if the TV script doesn't make perfect sense. It should look like alternating lines of dialogue, here is one such example of a few generated lines.

### Example generated script

>jerry: what about me?
>
>jerry: i don't have to wait.
>
>kramer:(to the sales table)
>
>elaine:(to jerry) hey, look at this, i'm a good doctor.
>
>newman:(to elaine) you think i have no idea of this...
>
>elaine: oh, you better take the phone, and he was a little nervous.
>
>kramer:(to the phone) hey, hey, jerry, i don't want to be a little bit.(to kramer and jerry) you can't.
>
>jerry: oh, yeah. i don't even know, i know.
>
>jerry:(to the phone) oh, i know.
>
>kramer:(laughing) you know...(to jerry) you don't know.

You can see that there are multiple characters that say (somewhat) complete sentences, but it doesn't have to be perfect! It takes quite a while to get good results, and often, you'll have to use a smaller vocabulary (and discard uncommon words), or get more data.  The Seinfeld dataset is about 3.4 MB, which is big enough for our purposes; for script generation you'll want more than 1 MB of text, generally. 

# Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_tv_script_generation.ipynb" and save another copy as an HTML file by clicking "File" -> "Download as.."->"html". Include the "helper.py" and "problem_unittests.py" files in your submission. Once you download these files, compress them into one zip file for submission.